### PR TITLE
feat: Add Settings to the App

### LINKED ISSUE #121: Add settings to the app ( as present in Event Brite )
In settings we can add the following things -

*How to use the app
*Version of app
*Rate Us ( Will be directed to the play store to rate the app )

Working on it

### PR DESCRIPTION
**Pull request by dreadpool2**
Fixes #121 

In settings->
* Added send suggestion to eventyay email
* For now "Rate us" will open organizer app in play store 
* Version name shown

videotogif_2018 06 18_16 04 56

### CODE REVIEW

File: app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        setHasOptionsMenu(true)
+
+        //Set Build Version
+        prefScreen.findPreference(resources.getString(R.string.key_version)).title = "Version " + BuildConfig.VERSION_NAME
+    }
+
+    override fun onPreferenceTreeClick(preference: Preference?): Boolean {
+        if (preference?.key == resources.getString(R.string.key_rating)) {
+            //Open Orga app in play store
+            startOrgaAppPlayStore("org.fossasia.eventyay")
```
Comment (nikit19) [inline app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt]: Why orga app ?

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -47,4 +48,19 @@
     <string name="order_range">Order Range</string>
     <string name="tickets_available">Tickets Available</string>
 
+    <!--suggestion mail details-->
+    <string name="emailInfo">List out your proposed Bug(s)/Feature(s)</string>
+    <string name="emailSubject">Improvement Suggestion for Open Event General App</string>
+    <string name="testEmailId">open.test24@gmail.com</string>
```
Comment (nikit19) [inline app/src/main/res/values/strings.xml]: Why this email? eventyay has its own email

**Comment by dreadpool2** [comment]
@iamareebjamal Should I make a logout section too ? Like in eventbrite:-
screenshot_20180618-161439

File: app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    private fun startOrgaAppPlayStore(packageName: String) {
+        try {
+            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + packageName)))
+        } catch (error: ActivityNotFoundException) {
+            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=" + packageName)))
+        }
+    }
+
+    private fun sendToTestEmail() {
```
Comment (nikit19) [inline app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt]: I don't think suggestions should be taken in an email. Eventbrite takes suggestions in a google form

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt]: Yeah! Then should I create a google form and use it ?

File: app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt (hunk 1 above)
Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt]: For now I kept orga app's package name as it is available in the play store. Should I change it to our package name then ?

File: app/src/main/res/values/strings.xml (hunk 2 above)
Comment (dreadpool2) [reply app/src/main/res/values/strings.xml]: Wasn't this intended for reporting bugs or features in the app? I thought that it would be better to receive emails directly about the bugs to the developers mail

Comment (nikit19) [reply app/src/main/res/values/strings.xml]: Please  go through reviews of other PRs as well https://github.com/fossasia/open-event-android/pull/150
and to know why this email was made you can have a look at the conversation here https://gitter.im/fossasia/open-event-android

Comment (dreadpool2) [reply app/src/main/res/values/strings.xml]: @nikit19 Bad misunderstanding! Thanks :+1:

**Comment by dreadpool2** [comment]
@iamareebjamal Should I make the logout section?

**Comment by iamareebjamal** [comment]
Yes

**Comment by dreadpool2** [comment]
Okay :+1:

**Comment by dreadpool2** [comment]
@iamareebjamal Please review :+1: 
In the eventbrite app suggestions are given in form of googleform, but here the user is sending a mail. What should I do ?

**Comment by dreadpool2** [comment]
@iamareebjamal Please Review :+1:

**Comment by iamareebjamal** [comment]
@nikit19 @simarsingh24 Please review

File: app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt (hunk 1 above)
Comment (nikit19) [reply app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt]: Yes

File: app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt (hunk 3 above)
Comment (nikit19) [reply app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt]: Yes

File: app/build.gradle [hunk 4]
```diff
@@ -66,7 +67,6 @@ dependencies {
     //Timber
     implementation 'com.jakewharton.timber:timber:4.7.0'
     implementation 'com.jakewharton.threetenabp:threetenabp:1.1.0'
-
```
Comment (nikit19) [inline app/build.gradle]: Not needed

File: app/src/main/java/org/fossasia/openevent/general/auth/ProfileFragment.kt [hunk 5]
```diff
@@ -35,7 +38,6 @@ class ProfileFragment : Fragment() {
         rootView = inflater.inflate(R.layout.fragment_profile, container, false)
 
         setHasOptionsMenu(true)
-
```
Comment (nikit19) [inline app/src/main/java/org/fossasia/openevent/general/auth/ProfileFragment.kt]: Not needed

File: app/src/main/java/org/fossasia/openevent/general/social/SocialLinksFragment.kt [hunk 6]
```diff
@@ -34,7 +34,6 @@ class SocialLinksFragment : Fragment() {
         rootView = inflater.inflate(R.layout.fragment_social_links, container, false)
 
         rootView.progressBarSocial.isIndeterminate = true
-
```
Comment (nikit19) [inline app/src/main/java/org/fossasia/openevent/general/social/SocialLinksFragment.kt]: Not needed

File: app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt [hunk 7]
```diff
@@ -24,7 +24,7 @@ class TicketsViewModel(private val ticketService: TicketService) : ViewModel() {
                 .observeOn(AndroidSchedulers.mainThread())
                 .doOnSubscribe({
                     progressTickets.value = true
-                }).doFinally({
+                }).doOnNext({
```
Comment (nikit19) [inline app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: Why?

File: app/src/main/java/org/fossasia/openevent/general/auth/ProfileFragment.kt [hunk 8]
```diff
@@ -16,11 +16,14 @@ import org.fossasia.openevent.general.R
...
 import org.koin.android.architecture.ext.viewModel
 import org.fossasia.openevent.general.AuthActivity
+import org.fossasia.openevent.general.settings.SettingsFragment
 
 class ProfileFragment : Fragment() {
     private val profileFragmentViewModel by viewModel<ProfileFragmentViewModel>()
 
     private lateinit var rootView: View
+    private var emailSettings: String? = null
+    private val EMAIL: String = "EMAIL"
```
Comment (simarsingh24) [inline app/src/main/java/org/fossasia/openevent/general/auth/ProfileFragment.kt]: const?

File: app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+import java.util.prefs.PreferenceChangeListener
+import org.fossasia.openevent.general.R
+import org.fossasia.openevent.general.BuildConfig
+import org.fossasia.openevent.general.MainActivity
+import org.koin.android.architecture.ext.viewModel
+import timber.log.Timber
+
+class SettingsFragment : PreferenceFragmentCompat(), PreferenceChangeListener {
+    private var email: String? = null
+    private val EMAIL: String = "EMAIL"
```
Comment (simarsingh24) [inline app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt]: same

File: app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt (hunk 7 above)
Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: Will change that :+1:

File: app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    override fun preferenceChange(evt: PreferenceChangeEvent?) {
+        preferenceChange(evt)
+    }
+
+    override fun onCreatePreferencesFix(savedInstanceState: Bundle?, rootKey: String?) {
+        // Load the preferences from an XML resource
+        setPreferencesFromResource(R.xml.settings, rootKey)
+        val prefScreen: PreferenceScreen = preferenceScreen
+
+        val activity = activity as? MainActivity
```
Comment (simarsingh24) [inline app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt]: is this necessary

File: app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+                activity?.onBackPressed()
+                true
+            }
+            else -> super.onOptionsItemSelected(item)
+        }
+    }
+
+    override fun onDestroyView() {
+        val activity = activity as? MainActivity
+        activity?.supportActionBar?.setDisplayHomeAsUpEnabled(false)
```
Comment (simarsingh24) [inline app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt]: why?

File: app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragmentViewModel.kt [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.schedulers.Schedulers
+import org.fossasia.openevent.general.auth.AuthService
+import timber.log.Timber
+
+class SettingsFragmentViewModel(private val authService: AuthService) : ViewModel() {
+
+    private val compositeDisposable = CompositeDisposable()
+
+    fun isLoggedIn() = authService.isLoggedIn()
```
Comment (simarsingh24) [inline app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragmentViewModel.kt]: cant we use directyl authservice.isloggedin from activivty

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragmentViewModel.kt]: I tried that, I couldn't though , thats why made a view model

File: app/src/main/java/org/fossasia/openevent/general/social/SocialLinksFragment.kt (hunk 6 above)
Comment (simarsingh24) [inline app/src/main/java/org/fossasia/openevent/general/social/SocialLinksFragment.kt]: unneeded change

File: app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt (hunk 7 above)
Comment (simarsingh24) [inline app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: why is this changed

File: app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt (hunk 11 above)
Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt]: The back home button will still be there even after I going back, We have implemented this in EventsFragmentDetails

**Comment by dreadpool2** [comment]
*Added link with google form
*[ Spreadsheet link ](https://docs.google.com/spreadsheet/d/1SzR75MBEVrTY1sDM3KAMm9wltiulDAp0QT5hv9eJkKM/edit#gid=1676755229)


videotogif_2018 06 20_01 35 34

**Comment by dreadpool2** [comment]
@iamareebjamal @nikit19 @simarsingh24 Please Review :+1:

**Comment by dreadpool2** [comment]
@iamareebjamal Please review this :+1:

File: app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+    override fun preferenceChange(evt: PreferenceChangeEvent?) {
+        preferenceChange(evt)
+    }
+
+    override fun onCreatePreferencesFix(savedInstanceState: Bundle?, rootKey: String?) {
+        // Load the preferences from an XML resource
+        setPreferencesFromResource(R.xml.settings, rootKey)
+
+        val activity = activity as? MainActivity
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt]: Never cast as specific activities, cast as generic action bar or support acitivity

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt]: We are taking this approach in every other fragment, I will create an issue :+1:

**Comment by dreadpool2** [comment]
PR Updated Please Review :+1:

File: app/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -56,4 +57,30 @@
...
+    <string name="profile_settings">PROFILE</string>
+    <string name="logged_in_account">Account</string>
+
+    <!--settings profile dialog-->
+    <string name="message">Are you sure you want to log out?</string>
+    <string name="cancel">Cancel</string>
+
+    <!--suggestion details-->
+    <string name="spread_sheet_link">https://docs.google.com/spreadsheets/d/1SzR75MBEVrTY1sDM3KAMm9wltiulDAp0QT5hv9eJkKM/edit#gid=1676755229</string>
+    <string name="suggestion_form_link">https://docs.google.com/forms/d/e/1FAIpQLSd7Y1T1xoXeYaAG_b6Tu1YYK-jZssoC5ltmQbkUX0kmDZaKYw/viewform</string>
```
Comment (iamareebjamal) [inline app/src/main/res/values/strings.xml]: Try not to use strings.xml for links. unneeded indirection

File: app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+        activity?.supportActionBar?.title = "Profile"
+        setHasOptionsMenu(false)
+        super.onDestroyView()
+    }
+
+    private fun showDialog() {
+        val builder = AlertDialog.Builder(activity)
+        builder.setMessage(resources.getString(R.string.message))
+                .setPositiveButton(resources.getString(R.string.logout)) { _, _ ->
+                    if(settingsViewModel.isLoggedIn()){
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt]: Space around symbols

File: app/src/main/res/values/strings.xml (hunk 14 above)
Comment (dreadpool2) [reply app/src/main/res/values/strings.xml]: @iamareebjamal The spreadsheet link is not being used in the app, but as the sheet collects the responses, it is required. Should I comment the link in SettingsFragment?

Comment (nikit19) [reply app/src/main/res/values/strings.xml]: why don't you add the link in the readme

Comment (dreadpool2) [reply app/src/main/res/values/strings.xml]: Great! I will do that :+1:

File: app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt (hunk 15 above)
Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/settings/SettingsFragment.kt]: PR Updated

File: app/src/main/res/values/strings.xml [hunk 16]
```diff
@@ -80,7 +80,6 @@
     <string name="cancel">Cancel</string>
 
     <!--suggestion details-->
-    <string name="spread_sheet_link">https://docs.google.com/spreadsheets/d/1SzR75MBEVrTY1sDM3KAMm9wltiulDAp0QT5hv9eJkKM/edit#gid=1676755229</string>
-    <string name="suggestion_form_link">https://docs.google.com/forms/d/e/1FAIpQLSd7Y1T1xoXeYaAG_b6Tu1YYK-jZssoC5ltmQbkUX0kmDZaKYw/viewform</string>
+    <string name="suggestion_form_link"></string>
```
Comment (iamareebjamal) [inline app/src/main/res/values/strings.xml]: Why empty entry

Comment (dreadpool2) [reply app/src/main/res/values/strings.xml]: Updated PR

File: app/src/main/res/values/strings.xml [hunk 17]
```diff
@@ -56,4 +57,26 @@
...
+    <string name="version_settings">Version</string>
+    <string name="key_version">version</string>
+    <string name="about_settings">ABOUT</string>
+    <string name="key_rating">rating</string>
+    <string name="rating_settings">Rate Us</string>
+    <string name="app_name_capital">OPEN EVENT ANDROID</string>
+    <string name="suggestion_settings">Suggest Improvement</string>
+    <string name="key_suggestion">suggestion</string>
+    <string name="key_profile">profile</string>
+    <string name="profile_settings">PROFILE</string>
```
Comment (iamareebjamal) [inline app/src/main/res/values/strings.xml]: No need for explicit capitalization

Comment (dreadpool2) [reply app/src/main/res/values/strings.xml]: Removed it :+1:

File: app/src/main/res/values/strings.xml [hunk 18]
```diff
@@ -65,14 +65,14 @@
...
+    <string name="about_settings">About</string>
     <string name="key_rating">rating</string>
     <string name="rating_settings">Rate Us</string>
-    <string name="app_name_capital">OPEN EVENT ANDROID</string>
+    <string name="app_name_capital">Open Event Android</string>
     <string name="suggestion_settings">Suggest Improvement</string>
     <string name="key_suggestion">suggestion</string>
     <string name="key_profile">profile</string>
-    <string name="profile_settings">PROFILE</string>
+    <string name="profile_settings">Profile</string>
```
[the remark this answers is not in the document]
Comment (dreadpool2) [reply app/src/main/res/values/strings.xml]: Yup :+1:  Updated PR :+1:

File: app/src/main/res/xml/settings.xml [hunk 19]
```diff
@@ -13,7 +13,7 @@
     </PreferenceCategory>
 
     <PreferenceCategory
-        android:title="@string/profile_settings" >
+        android:title="@string/Profile" >
```
Comment (iamareebjamal) [inline app/src/main/res/xml/settings.xml]: `profile`

Comment (dreadpool2) [reply app/src/main/res/xml/settings.xml]: Changed it :+1:
